### PR TITLE
Authentication

### PR DESCRIPTION
**Pull request by foolhardy21**
## Auth Pages

This PR has APIs and UI components for login and signup pages.

- Files to review:
   - [ ] src/reducers/signup.reducer.js
   - [ ] src/reducers/login.reducer.js
   - [ ] src/pages/Signup.jsx
   - [ ] src/pages/Login.jsx
   - [ ] src/contexts/auth.context.js
   - [ ] src/components/Signup/SignupForm.jsx
   - [ ] src/components/Login/LoginForm.jsx

- Use this link for preview - https://deploy-preview-2--remarkable-flan-74a807.netlify.app/signup

v-1
v-2

Thanks.

### CODE REVIEW

File: src/pages/Signup.jsx [hunk 1]
```diff
@@ -0,0 +1,19 @@
+import { SignupCard } from "../components/Signup"
+
+const Signup = () => {
+
+    return (
+        <div
+            style={{
```
Comment (hsnice16) [inline src/pages/Signup.jsx:7]: I think we should avoid using inline styling. So, see if you can make a class for this.

**Review by hsnice16 — APPROVED** [review]
Nice work. Just added a suggestion.

And also, I don't know if you are using any code formatting extension or not but I would suggest to use one. I think, the code was not formatted properly in most of files, like in loginform and signupform. So you can look into this.

**Comment by foolhardy21** [comment]
> Nice work. Just added a suggestion.
> 
> And also, I don't know if you are using any code formatting extension or not but I would suggest to use one. I think, the code was not formatted properly in most of files, like in loginform and signupform. So you can look into this.

thanks @hsnice16 . I'll include code formatting.

File: src/pages/Signup.jsx (hunk 1 above)
Comment (foolhardy21) [reply src/pages/Signup.jsx:7]: as I want the `min-height` to be `100vh` always, I thought of using inline style for its priority over class or id selectors.

File: src/components/Login/LoginForm.jsx [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+        if (isFormEmpty({ username, password })) {
+            showAlert(loginDispatch, 'form is empty', ALERT_TYPE_ERROR)
+        } else {
+            const response = await logInUser(username, password)
+            if (response.status === 200) {
+                showAlert(loginDispatch, 'logged in', ALERT_TYPE_SUCCESS)
+            } else if (response.status === 404) {
+                showAlert(loginDispatch, 'user not found', ALERT_TYPE_ERROR)
+            } else if (response.status === 401) {
+                showAlert(loginDispatch, 'wrong password', ALERT_TYPE_ERROR)
```
Comment (anik31) [inline src/components/Login/LoginForm.jsx:43]: appreciation - good that you're showing all types of error at login

File: src/utils/constants.util.js [hunk 3]
```diff
@@ -0,0 +1,14 @@
+export const ACTION_TOGGLE_PASSWORD_TYPE = 'TOGGLE_PASSWORD_TYPE'
+export const ACTION_UPDATE_USERNAME = 'UPDATE_USERNAME'
+export const ACTION_UPDATE_PASSWORD = 'UPDATE_PASSWORD'
+export const ACTION_UPDATE_ALERT = 'UPDATE_ALERT'
+export const ACTION_UPDATE_FIRST_NAME = 'UPDATE_FIRSTNAME'
+export const ACTION_UPDATE_LAST_NAME = 'UPDATE_LASTNAME'
+
+export const API_LOGIN = '/api/auth/login'
+export const API_SIGNUP = '/api/auth/signup'
+
```
Comment (anik31) [inline src/utils/constants.util.js:10]: appreciation - good that you're maintaining a file for constants

File: src/utils/form.util.js [hunk 4]
```diff
@@ -0,0 +1,6 @@
+/**
+ * this function checks if the form object has any empty values or not
+ * @param {Object.prototype} formObj - an object where keys denote the input fields in the form and the values are their respective values enetered by the user
+ * @returns {boolean} - true if the form is empty otherwise false 
+ */
+export const isFormEmpty = formObj => Object.values(formObj).some(val => val === '')
```
Comment (anik31) [inline src/utils/form.util.js:6]: appreciation - liked the use of js doc

**Review by anik31 — APPROVED** [review]
Great work @foolhardy21 

suggestion - you can add an option for test login

**Comment by foolhardy21** [comment]
> Great work @foolhardy21 
> 
> suggestion - you can add an option for test login

Sure @anik31, I'll add the feature.

**Pull request merged**